x-pack/plugins/reporting/server/export_types/csv_searchsource/generate_csv/generate_csv.ts

```diff
@@ -0,0 +1,250 @@
+    const { checkForFormulas, escapeValue, separator } = settings;
+
+    for (const dataTableRow of table.rows) {
+      if (this.cancellationToken.isCancelled()) {
+        break;
+      }
+
+      const columnMap = this.getColumnMap(fields, table);
+      const row = columnMap
+        .map((columnIndex) => {
+          const tableColumn = table.columns[columnIndex];
+
+          let cell = this.getFormatters(table)[tableColumn.id].convert(
+            dataTableRow[tableColumn.id]
+          );
+
+          // We have to strip singular array values out of their array wrapper,
+          // So that the value appears the visually the same as seen in Discover
+          // TODO move this to tabifyDocs
+          try {
+            const cellData = JSON.parse(cell);
+            if (Array.isArray(cellData) && cellData.length === 1) {
+              cell = cellData[0];
+            }
+          } catch (e) {
+            // allow parse error to be silent
+          }
+
+          return cell;
+        })
+        .map((cell) => {
+          if (checkForFormulas && cellHasFormulas(cell)) {
+            this.csvContainsFormulas = true; // set warning if cell value has a formula
+          }
+          return escapeValue(cell);
+        });
+
+      if (!builder.tryAppend(row.map(escapeValue).join(separator) + '\n')) {
+        this.logger.warn('max Size Reached');
+        this.maxSizeReached = true;
+        if (this.cancellationToken) {
+          this.cancellationToken.cancel();
+        }
+        break;
+      }
+
+      this.csvRowCount++;
+    }
+  }
+
+  public async generateData(): Promise<SavedSearchGeneratorResult> {
+    const [settings, searchSource] = await Promise.all([
+      getExportSettings(this.job.browserTimezone, this.uiSettingsClient, this.config, this.logger),
+      this.searchSourceService.create(this.job.searchSource),
+    ]);
+
+    const index = searchSource.getField('index');
+    const fields = searchSource.getField('fields');
+    const { maxSizeBytes, bom, escapeFormulaValues } = settings;
+
+    const builder = new MaxSizeStringBuilder(byteSizeValueToNumber(maxSizeBytes), bom);
+
+    const recordsAtATime = settings.scrollSize;
+    let currentRecord = -1;
+    let totalRecords = 0;
+    let first = true;
+    let lastSortId: EsQuerySearchAfter | undefined;
+    const warnings: string[] = [];
+    searchSource.setField('size', recordsAtATime);
+
+    while (currentRecord < totalRecords) {
+      if (lastSortId) {
+        searchSource.setField('searchAfter', lastSortId);
+      }
+
+      const results = await searchSource.fetch(); // ignores the selected fields option
+      totalRecords = results.hits.total;
+      if (totalRecords == null) {
+        throw new Error('Expected total number of records in the search response');
+      }
+      if (results.hits.hits.length > 0) {
+        lastSortId = results.hits.hits[results.hits.hits.length - 1].sort as EsQuerySearchAfter; // BUG
```

Comment:
BUG: The `sort` property is not guaranteed to be a unique value for all the data. Documents are getting lost in between paging the multiple searches to get all the data.

```
[2021-02-02T23:42:31.727Z][DEBUG][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] Search results: 1266
[2021-02-02T23:42:31.727Z][DEBUG][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] Building CSV header row...
[2021-02-02T23:42:31.728Z][DEBUG][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] Building 500 CSV data rows...
[2021-02-02T23:42:31.827Z][DEBUG][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] Building 500 CSV data rows...
[2021-02-02T23:42:31.893Z][DEBUG][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] Building 265 CSV data rows...
[2021-02-02T23:42:31.925Z][DEBUG][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] Finished generating. Total size in bytes: 80203. Row count: 1265.
[2021-02-02T23:42:31.926Z][WARN ][plugins.reporting.csv_searchsource_immediate.csv_searchsource_immediate.execute-job] CSV may contain formulas whose values have been escaped
[2021-02-02T23:42:31.926Z][INFO ][plugins.reporting.csv_searchsource_immediate] Job output size: 80203 bytes
```
^^ The last search results should have been 266 documents